# CPSC 330 - Applied Machine Learning 

## Homework 7: Clustering and recommender systems
### Associated lectures: Lectures 14 and 15

**Due date: Wednesday, March 22, 11:59pm**


In [1]:
import os

%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.cluster import DBSCAN, KMeans
from sklearn.feature_extraction.text import CountVectorizer

pd.set_option("display.max_colwidth", 0)

import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

<br><br><br><br>

## Instructions 
<hr>
rubric={points:2}

Follow the [homework submission instructions](https://github.com/UBC-CS/cpsc330-2022W2/blob/main/docs/homework_instructions.md). 

**You may work on this homework in a group and submit your assignment as a group.** Below are some instructions on working as a group.  
- The maximum group size is 2. 
- Use group work as an opportunity to collaborate and learn new things from each other. 
- Be respectful to each other and make sure you understand all the concepts in the assignment well. 
- It's your responsibility to make sure that the assignment is submitted by one of the group members before the deadline. 
- You can find the instructions on how to do group submission on Gradescope [here](https://help.gradescope.com/article/m5qz2xsnjy-student-add-group-members).

<br><br><br><br>

## Exercise 1: Document clustering toy example <a name="1"></a>
<hr>

In lecture 14, we looked at a popular application of clustering: customer segmentation. In this homework, we will work on a toy example of another popular application: [**document clustering**](https://en.wikipedia.org/wiki/Document_clustering). A large amount of unlabeled text data is available out there (e.g., news, recipes, online Q&A), and clustering is a commonly used technique to organize this data in a meaningful way. 

In this exercise, we will create a toy dataset with sentences from Wikipedia articles and cluster these sentences. 

### 1.1 Sample sentences from Wikipedia articles
rubric={points:2}

The code below extracts first sentences of Wikipedia articles on a set of queries. You will need the `wikipedia` package installed in the course environment to run the code below. 

```
conda activate cpsc330
conda install -c conda-forge wikipedia
```

You also need `nltk` library in the course environment. 

```
conda install -c anaconda nltk 
```        

**Your tasks:**

Run the code below and answer the following question. 

1. Given this dataset, how many clusters would you expect a clustering algorithm to identify? How would you manually label these clusters?   

> *Note: Feel free to experiment with queries of your choice. But stick to the provided list for the final submission so that it's easier for the TAs when they grade your submission.*


In [2]:
import nltk
from nltk.tokenize import sent_tokenize, word_tokenize
nltk.download("punkt")

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\blais\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [17]:
import wikipedia

queries = [
    "Mount Everests",
    "Raspberry",
    "Mount Denali",
    "Arithmetic",
    "Topology",
    "Baseball",
    "Hockey",
    "Mango_fruit",
    "Mount Kenya",
    "Football"
]

wiki_dict = {"wiki query": [], "text": [], "n_words": []}
for i in range(len(queries)):
    sent = sent_tokenize(wikipedia.page(queries[i]).content)[0]
    wiki_dict["text"].append(sent)
    wiki_dict["n_words"].append(len(word_tokenize(sent)))
    wiki_dict["wiki query"].append(queries[i])

wiki_df = pd.DataFrame(wiki_dict)
wiki_df

wiki query  \
0  Mount Everests   
1  Raspberry        
2  Mount Denali     
3  Arithmetic       
4  Topology         
5  Baseball         
6  Hockey           
7  Mango_fruit      
8  Mount Kenya      
9  Football         

                                                                                                                                                                                                                                                                                                                                                                                                                                                               text  \
0  Mount Everest (Nepali: सगरमाथा, romanized: Sagarmāthā; Tibetan: Chomolungma ཇོ་མོ་གླང་མ; Chinese: 珠穆朗玛峰; pinyin: Zhūmùlǎngmǎ Fēng) is Earth's highest mountain above sea level, located in the Mahalangur Himal sub-range of the Himalayas.                                                                                                                                                                                                                        
1  The raspberry is the edible fruit of a multitude of plant species in the genus Rubus of the rose family, most of which are in the subgenus Idaeobatus.                                                                                                                                                                                                                                                                                                             
2  Denali (; also known as Mount McKinley, its former official name) is the highest mountain peak in North America, with a summit elevation of 20,310 feet (6,190 m) above sea level.                                                                                                                                                                                                                                                                                 
3  Arithmetic (from Ancient Greek  ἀριθμός (arithmós) 'number', and  τική [τέχνη] (tikḗ [tékhnē]) 'art, craft') is an elementary part of mathematics that consists of the study of the properties of the traditional operations on numbers—addition, subtraction, multiplication, division, exponentiation, and extraction of roots.                                                                                                                                  
4  In mathematics, topology (from the Greek words τόπος, 'place, location', and λόγος, 'study') is concerned with the properties of a geometric object that are preserved under continuous deformations, such as stretching, twisting, crumpling, and bending; that is, without closing holes, opening holes, tearing, gluing, or passing through itself.                                                                                                             
5  Basketball is a team sport in which two teams, most commonly of five players each, opposing one another on a rectangular court, compete with the primary objective of shooting a basketball (approximately 9.4 inches (24 cm) in diameter) through the defender's hoop (a basket 18 inches (46 cm) in diameter mounted 10 feet (3.048 m) high to a backboard at each end of the court), while preventing the opposing team from shooting through their own hoop.   
6  Hockey is a term used to denote a family of various types of both summer and winter team sports which originated on either an outdoor field, sheet of ice, or dry floor such as in a gymnasium.                                                                                                                                                                                                                                                                    
7  A mango is an edible stone fruit produced by the tropical tree Mangifera indica.                                                    

# **Solution**
Based on this dataset, I would expect a good clustering algorithm to identify 4 clusters namely Mathematics, Mountains, Fruits and Sport.
- Mathematics
  - Arithmetic
  - Topology
- Mountains
  - Mount Everest
  - Mount Denali
  - Mount Kenya
- Fruits
  - Raspberry
  - mango_fruit
- Sport
  - Baseball
  - Hockey
  - Football 

<br><br>

### 1.2 `KMeans` with bag-of-words representation 
rubric={points:10}

We have seen that before we pass text to machine learning models, we need to encode it into a numeric representation. So let's encode our toy dataset above (`wiki_df`) to a numeric representation. 

First, let's try our good old friend: bag-of-words representation. The code below creates dense bag-of-words representation of Wikipedia sentences from question 1.1 using a [`CountVectorizer`](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html). 

**Your tasks:**

Run the code below and answer the following questions. 

1. Run `KMeans` clustering on the transformed data (`bow_sents`) with K = the number of clusters you identified in question 1.1. Examine clustering labels assigned by `KMeans`.
2. Repeat after modifying the `CountVectorizer` to ignore words appearing in only one sentence (or document, one row of `wiki_df`).
3. Keeping the new `CountVectorizer`, examine clustering labels assigned by `KMeans` under an off-by-1 value of K.
4. Briefly describe and analyze the behavoir of `KMeans`. Is `KMeans` doing a reasonable job in clustering the sentences? 

> You can access cluster label assignments using `labels_` attribute of the clustering object. 

In [18]:
vec = CountVectorizer(stop_words='english')
bow_sents = vec.fit_transform(wiki_df["text"]).todense()
bow_df = pd.DataFrame(
    data=bow_sents, columns=vec.get_feature_names_out(), index=wiki_df.index
)
bow_df

048  10  18  190  20  24  310  46  addition  africa  ...  winter  words  \
0  0    0   0   0    0   0   0    0   0         0       ...  0       0       
1  0    0   0   0    0   0   0    0   0         0       ...  0       0       
2  0    0   0   1    1   0   1    0   0         0       ...  0       0       
3  0    0   0   0    0   0   0    0   1         0       ...  0       0       
4  0    0   0   0    0   0   0    0   0         0       ...  0       1       
5  1    1   1   0    0   1   0    1   0         0       ...  0       0       
6  0    0   0   0    0   0   0    0   0         0       ...  1       0       
7  0    0   0   0    0   0   0    0   0         0       ...  0       0       
8  0    0   0   0    0   0   0    0   0         1       ...  0       0       
9  0    0   0   0    0   0   0    0   0         0       ...  0       0       

   zhūmùlǎngmǎ  λόγος  τέχνη  τική  τόπος  सगरम  ἀριθμός  珠穆朗玛峰  
0  1            0      0      0     0      1     0        1      
1  0            0      0      0     0      0     0        0      
2  0            0      0      0     0      0     0        0      
3  0            0      1      1     0      0     1        0      
4  0            1      0      0     1      0     0        0      
5  0            0      0      0     0      0     0        0      
6  0            0      0      0     0      0     0        0      
7  0            0      0      0     0      0     0        0      
8  0            0      0      0     0      0     0        0      
9  0            0      0      0     0      0     0        0      

[10 rows x 169 columns]

**Solution**

In [20]:
wiki_querries = wiki_df[["wiki query"]].copy(deep=True)
kmeans = KMeans(n_clusters=4, random_state=42);
kmeans.fit(bow_df);
# cluster_preds = kmeans.predict(bow_df)
wiki_querries["Cluster"]=kmeans.labels_ #cluster_preds
wiki_querries.sort_values("Cluster").style.set_caption("Kmeans on bow_df")
# print(cluster_preds)

In [21]:
# Creating a new Count Vectorizer Instantiation
vec_min_df = CountVectorizer(stop_words='english', min_df=2)

bow_sents_min_df = vec_min_df.fit_transform(wiki_df["text"]).todense()
bow_min_df = pd.DataFrame(
    data=bow_sents_min_df, columns=vec_min_df.get_feature_names_out(), index=wiki_df.index
)
bow_min_df

wiki_querries = wiki_df[["wiki query"]].copy(deep=True)
kmeans_min_df = KMeans(n_clusters=4, random_state=42);
kmeans_min_df.fit(bow_min_df);
# cluster_preds = kmeans_min_df.predict(bow_min_df)
wiki_querries["Cluster"]= kmeans_min_df.labels_#cluster_preds
wiki_querries.sort_values("Cluster").style.set_caption("Adjusted count vectorisor")

In [23]:
# Kmeans by k = 6
bow_min_df
wiki_querries = wiki_df[["wiki query"]].copy(deep=True)
off_by_1_kmeans = KMeans(n_clusters=5, random_state=42)
off_by_1_kmeans.fit(bow_min_df)
wiki_querries["Cluster"]= off_by_1_kmeans.labels_#cluster_preds
wiki_querries.sort_values("Cluster").style.set_caption("Of by 1 value of k")



<br><br>

### 1.3 Sentence embedding representation
rubric={points:10}

Bag-of-words representation is limited in that it does not take into account word ordering and context. There are other richer representations of text, and we are going to use one such representation in this lab. 

The code below creates an alternative and a more expressive representation of sentences. We will call it *sentence embedding representation*. We'll use [sentence transformer](https://www.sbert.net/index.html) to extract these representations. At this point it's enough to know that this is an alternative representation of text which usually works better than simple bag-of-words representation. We will talk a bit more about embedding representations next week. You need to install `sentence-transformers` in the course conda environment to run the code below. 

```conda install -c conda-forge sentence-transformers```

**Your tasks:**

Run the code below and answer the following questions. 

1. How many dimensions (features associated with each example) are present in this representation? 
2. Run `KMeans` clustering with sentence embedding representation of text (`emb_sents`) and examine cluster labels. 
3. How well the sentences are clustered together? 

In [24]:
from sentence_transformers import SentenceTransformer

embedder = SentenceTransformer("paraphrase-distilroberta-base-v1")

c:\Users\blais\Anaconda3\envs\DeepLearningEnv\lib\site-packages\tqdm\auto.py:22: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [43]:
emb_sents = embedder.encode(wiki_df["text"])
emb_sent_df = pd.DataFrame(emb_sents, index=wiki_df.index)
emb_sent_df

0         1         2         3         4         5         6  \
0 -0.532166  0.219127  0.029919 -0.407121 -0.616050  0.339105 -0.369831   
1 -0.257144  0.279162  0.222628  0.190509  0.057312  0.215432  0.124739   
2 -0.153511  0.308003  0.042989 -0.000984 -0.228809  0.145077 -0.189960   
3 -0.182559  0.174978 -0.142932  0.508398 -0.102792  0.314821 -0.028197   
4  0.002731  0.032421  0.043655  0.501372  0.442546  0.157427  0.082318   
5 -0.042878  0.029167 -0.187499 -0.020230 -0.260429 -0.120475  0.067831   
6 -0.071380  0.096587  0.055705 -0.525644  0.580350  0.096592  0.441485   
7 -0.141320  0.030475  0.261100 -0.219024 -0.277812  0.177240  0.191266   
8 -0.291509  0.554150  0.115631 -0.188634 -0.074476 -0.277101 -0.000275   
9 -0.071756  0.133156 -0.071762  0.306060  0.358961 -0.061442  0.167584   

          7         8         9  ...       758       759       760       761  \
0 -0.268620  0.284336  0.154874  ...  0.049629 -0.063531  0.193070  0.345019   
1  0.256541 -0.093520  0.304061  ... -0.350747 -0.103448  0.188898  0.002680   
2  0.006522  0.346580  0.009805  ...  0.345482  0.149473 -0.128204  0.069987   
3  0.235509  0.293719  0.091493  ...  0.153091 -0.220330  0.126607  0.091784   
4  0.287892  0.322357  0.356958  ...  0.267656 -0.337920  0.343272  0.429253   
5  0.296153 -0.020178  0.003125  ...  0.178184 -0.003942  0.331448 -0.395468   
6  0.257062 -0.054260  0.220843  ...  0.457613 -0.331445 -0.066380 -0.019425   
7  0.252725 -0.035201  0.198338  ...  0.104466 -0.177960  0.088619  0.289035   
8  0.076826  0.199135  0.033997  ...  0.055575  0.047854  0.214838 -0.086142   
9  0.364614  0.323540  0.141398  ...  0.404115  0.000236  0.177493  0.016162   

        762       763       764       765       766       767  
0  0.111516 -0.018490  0.062622 -0.105441  0.084189  0.097068  
1  0.364525  0.270560  0.617304  0.565283  0.068011  0.210108  
2 -0.049316 -0.122242 -0.520354  0.052404  0.277885 -0.082411  
3  0.087273  0.279831 -0.328509  0.157603  0.439282  0.204193  
4 -0.042079  0.466611 -0.574507  0.004958  0.390041  0.150983  
5  0.025013  0.082307 -0.561699  0.560587  0.141633 -0.373937  
6 -0.147649  0.034530 -0.095109  0.304422  0.327636 -0.197769  
7 -0.089657  0.246049 -0.060229  0.500508 -0.045897 -0.046336  
8  0.453633  0.021888 -0.143639  0.167839 -0.151438 -0.076444  
9  0.232335  0.260556  0.168149  0.483962  0.121122  0.196881  

[10 rows x 768 columns]

In [44]:
dimensions = emb_sent_df.shape
print(f"There are {dimensions[1]} features per example")

There are 768 features per example


In [45]:
wiki_df.columns

Index(['wiki query', 'text', 'n_words'], dtype='object')

In [46]:
emb_kmeans = KMeans(n_clusters=4)
emb_kmeans.fit(emb_sent_df)

wiki_querries = wiki_df[['wiki query']].copy(deep = True)

emb_clusters = emb_kmeans.labels_
wiki_querries["Cluster"] = emb_clusters
wiki_querries.sort_values("Cluster")

wiki query  Cluster
5  Baseball        0      
6  Hockey          0      
9  Football        0      
1  Raspberry       1      
7  Mango_fruit     1      
0  Mount Everests  2      
2  Mount Denali    2      
8  Mount Kenya     2      
3  Arithmetic      3      
4  Topology        3

<br><br>

### 1.4 DBSCAN with cosine distance  
rubric={points:10}

Let's try `DBSCAN` on our toy dataset. K-Means is kind of bound to the Euclidean distance because it is based on the notion of means. With `DBSCAN` we can try different distance metrics. In the context of text (sparse data), [cosine similarities](https://scikit-learn.org/stable/modules/metrics.html#cosine-similarity) or cosine distances tend to work better. Given vectors $u$ and $v$, the **cosine distance** between the vectors is defined as: 

$$distance_{cosine}(u,v) = 1 - (\frac{u \cdot v}{\left\lVert u\right\rVert_2 \left\lVert v\right\rVert_2})$$

In this exercise, you'll use DBSCAN with cosine distances. 

**Your tasks**

1. Use DBSCAN to cluster our toy data using sentence embedding representation (`emb_sents`) and `metric='cosine'`. 
2. Briefly comment on the number of clusters identified and the cluster assignment given by the algorithm.

> *Note: You will also have to set appropriate values for the hyperparamters `eps` and `min_samples` to get meaningful clusters, as default values for these hyperparameters won't work on this toy dataset. In order to set appropriate value for `eps`, you may want to examine the distances given by [sklearn's `cosine_distance`](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.paired_cosine_distances.html).*

In [47]:
from sklearn.metrics.pairwise import cosine_similarity
cosine_similarity(emb_sents)

array([[0.9999999 , 0.06866539, 0.5756489 , 0.17209145, 0.1298811 ,
        0.07449248, 0.04888136, 0.15815566, 0.5905613 , 0.00804082],
       [0.06866539, 1.0000002 , 0.09353441, 0.15299201, 0.13539034,
        0.1303733 , 0.09676598, 0.47787338, 0.10128872, 0.16074818],
       [0.5756489 , 0.09353441, 1.0000002 , 0.09290385, 0.14520553,
        0.14353049, 0.18436618, 0.14655231, 0.54304373, 0.0955447 ],
       [0.17209145, 0.15299201, 0.09290385, 1.0000004 , 0.61889696,
        0.22745517, 0.18740621, 0.09766782, 0.11001416, 0.18310389],
       [0.1298811 , 0.13539034, 0.14520553, 0.61889696, 0.99999994,
        0.24488428, 0.24643438, 0.15702598, 0.04495554, 0.20752472],
       [0.07449248, 0.1303733 , 0.14353049, 0.22745517, 0.24488428,
        0.99999994, 0.40370578, 0.04856509, 0.16898268, 0.47169632],
       [0.04888136, 0.09676598, 0.18436618, 0.18740621, 0.24643438,
        0.40370578, 1.        , 0.08395086, 0.04546799, 0.4738096 ],
       [0.15815566, 0.47787338, 0.1465523

In [49]:
db_scan = DBSCAN(metric="cosine", min_samples=2, eps=0.1)
db_scan.fit(emb_sents)

wiki_querries = wiki_df[['wiki query']].copy(deep = True)

db_clusters = db_scan.labels_#.labels_
wiki_querries["DB Cluster"] = emb_clusters
wiki_querries.sort_values("DB Cluster")

wiki query  DB Cluster
5  Baseball        0         
6  Hockey          0         
9  Football        0         
1  Raspberry       1         
7  Mango_fruit     1         
0  Mount Everests  2         
2  Mount Denali    2         
8  Mount Kenya     2         
3  Arithmetic      3         
4  Topology        3

<br><br>

### 1.5 Visualizing clusters 
rubric={points:5}

One thing we could do with unlabeled data is visualizing it. That said, our data is high dimensional (each example is represented with 768 dimensions) and high-dimensional data is hard to visualize. One way to visualize high-dimensional data is applying dimensionality reduction to get the most important (2 or 3) components of the dataset and visualizing this low-dimensional data. 

Given data as a `numpy` array and cluster assignments, the `plot_pca_clusters` function below transforms the given data by applying dimensionality reduction and plots the transformed data into corresponding clusters. 

> *Note: At this point we are using this function only for visualization and you are not expected to understand the PCA part. Feel free to modify the function as you see fit.*

**Your tasks:**

1. Call the function `plot_pca_clusters` to visualize the clusters created by the three models above:
    - KMeans with bag-of-words representation 
    - KMeans with sentence embedding representation 
    - DBSCAN with sentence embedding representation 

In [50]:
from sklearn.decomposition import PCA # Obtain the principal components

def plot_pca_clusters(
    data,
    cluster_labels,
    raw_sents=wiki_df["text"],
    show_labels=False,
    size=100,
    title="PCA visualization",
):
    """
    Carry out dimensionality reduction using PCA and plot 2-dimensional clusters.

    Parameters
    -----------
    data : numpy array
        data as a numpy array
    cluster_labels : list
        cluster labels for each row in the dataset
    raw_sents : list
        the original raw sentences for labeling datapoints
    show_labels : boolean
        whether you want to show labels for points or not (default: False)
    size : int
        size of points in the scatterplot
    title : str
        title for the visualization plot

    Returns
    -----------
    None. Shows the clusters.
    """

    pca = PCA(n_components=2)
    principal_comp = pca.fit_transform(data)
    pca_df = pd.DataFrame(data=principal_comp, columns=["pca1", "pca2"])
    pca_df["cluster"] = cluster_labels

    plt.figure(figsize=(10, 7))
    plt.title(title)
    ax = sns.scatterplot(
        x="pca1", y="pca2", hue="cluster", data=pca_df, palette="tab10", s=size
    )

    x = pca_df["pca1"].tolist()
    y = pca_df["pca2"].tolist()
    if show_labels:
        for i, txt in enumerate(raw_sents):
            plt.annotate(" ".join(txt.split()[:10]), (x[i], y[i]))
        ax.legend(loc="upper right")

    plt.show()

<br><br><br><br>

## Exercise 2: Movie recommendations
<hr>

Let's build simple movie recommendation systems using the [MovieLens dataset](https://www.kaggle.com/prajitdatta/movielens-100k-dataset/data). The original source of the data is [here](https://grouplens.org/datasets/movielens/), and the structure of the data is described in the [README](http://files.grouplens.org/datasets/movielens/ml-latest-small-README.html) that comes with it. The code below reads the data as a CSV assuming that it's under `data/ml-100k/` directory under your lab folder. 

In [ ]:
r_cols = ["user_id", "movie_id", "rating", "timestamp"]
ratings = pd.read_csv(
    os.path.join("data", "ml-100k", "u.data"),
    sep="\t",
    names=r_cols,
    encoding="latin-1",
)
ratings.head()

In [ ]:
# We'll be using these keys later in the starter code
user_key = "user_id"
item_key = "movie_id"

### 2.1 Terminology
rubric={points:6}

Here is some notation we will be using in this lab. 

**Constants**:

 - $N$: the number of users, indexed by $n$
 - $M$: the number of movies, indexed by $m$
 - $\mathcal{R}$: the set of indices $(n,m)$ where we have ratings in the utility matrix $Y$
    - Thus $|\mathcal{R}|$ is the total number of ratings
 - $k$: the number of latent dimensions we use in collaborative filtering
 
**The data**:

 - $Y$: the utility matrix containing ratings, with a lot of missing entries
 - $Z$: a matrix whose rows $z_m$ represent the features for movie $m$ (size $M\times d$).
 - `train_mat` and `valid_mat`: Utility matrices for train and validation sets, respectively
 
    
**Your tasks:**    

1. What are the values of $N$ and $M$ in movie ratings data?  
2. What would be the shape of the dense utility matrix $Y$? 
3. What would be the fraction of observed ratings in the utility matrix $Y$? 

In [ ]:
N = None
M = None

<br><br>

### 2.2 Splitting the data
rubric={points:5}

**Your tasks:**

1. Split the ratings data with `test_size=0.2` and `random_state=42`. 

<br><br>

### 2.3 Utility matrix 
rubric={points:10}

**Your tasks**
1. Create utility matrices for train and validation sets (similar to how we did in the class). 
2. Briefly explain the difference between the train and validation utility matrices. 

> You may use the code from lecture notes with appropriate attributions.  

> You won't do it in real life but since our dataset is not that big, create a dense utility matrix in this assignment. You are welcome to try sparse matrix but then you may have to change some started code provided in the later exercises.  

In [ ]:
user_mapper = dict(zip(np.unique(ratings[user_key]), list(range(N))))
item_mapper = dict(zip(np.unique(ratings[item_key]), list(range(M))))
user_inverse_mapper = dict(zip(list(range(N)), np.unique(ratings[user_key])))
item_inverse_mapper = dict(zip(list(range(M)), np.unique(ratings[item_key])))

In [ ]:
train_mat = None
valid_mat = None

<br><br>

### 2.4 Evaluation and baseline
rubric={points:4}

To compare different models you build in this homework, let's write a couple of functions for evaluation. 
- The `error` function returns RMSE.
- The `evaluate` function prints the train and validation RMSEs. 

**Your task:**

1. Briefly explain what exactly we are comparing to evaluate recommender systems. 
2. Implement the global average baseline, where you predict everything as the global average rating. What's the RMSE of the global average baseline?

In [ ]:
def error(Y1, Y2):
    """
    Returns the root mean squared error (RMSE).
    """
    return np.sqrt(np.nanmean((Y1 - Y2) ** 2))


def evaluate(pred_Y, train_mat, valid_mat, model_name="Global average"):
    print("%s train RMSE: %0.2f" % (model_name, error(pred_Y, train_mat)))
    print("%s valid RMSE: %0.2f" % (model_name, error(pred_Y, valid_mat)))

<br><br>

### (Optional) 2.5 $k$-nearest neighbours imputation
rubric={points:1}

**Your tasks:**

Try [KNNImputer](https://scikit-learn.org/stable/modules/generated/sklearn.impute.KNNImputer.html) to fill in the missing entries. Discuss your observations. 

<br><br>

### 2.6 Use collaborative filtering with the `surprise` package
rubric={points:6}

Use the [`surprise`](https://surprise.readthedocs.io/en/stable/) package which has implementation of SVD algorithm for collaborative filtering. You can install it as follows in your conda environment. 

```
>> conda activate cpsc330
>> conda install -c conda-forge scikit-surprise
or 
>> pip install scikit-surprise
```

**Your tasks:**

1. Carry out cross-validation using SVD algorithm in the package, similar to how we did it in the lecture on Jester dataset. Report mean RMSE and compare it with global baseline. 

<br><br>

### 2.7 Clustering based recommendation system?
rubric={points:2}

How would you apply `K-Means` clustering to build a recommendation systems? What could be challenging with this approach?

<br><br><br><br>

## Submission instructions 

**PLEASE READ:** When you are ready to submit your assignment do the following:

1. Run all cells in your notebook to make sure there are no errors by doing `Kernel -> Restart Kernel and Clear All Outputs` and then `Run -> Run All Cells`. 
2. Notebooks with cell execution numbers out of order or not starting from “1” will have marks deducted. Notebooks without the output displayed may not be graded at all (because we need to see the output in order to grade your work).
3. Upload the assignment using Gradescope's drag and drop tool. Check out this [Gradescope Student Guide](https://lthub.ubc.ca/guides/gradescope-student-guide/) if you need help with Gradescope submission. 